### PR TITLE
LoadMap in Game

### PR DESCRIPTION
**Pull request by urban119**
This allow load map as specified point (x,y). 
Example:
Game.loadMap("data/world/castle.otbm", 662, 371);
Map Castle.otbm is position (x=1,y=1)
https://youtu.be/gNEJnYDhcT8?t=46s

Old:
Game.loadMap("data/world/castle.otbm");

Regards
Urban

### CODE REVIEW

File: src/iomap.cpp [hunk 1]
```diff
@@ -198,9 +198,22 @@ bool IOMap::loadMap(Map* map, const std::string& identifier)
...
+			uint16_t base_y = 0;
+			uint16_t offset_x = 0;
+			uint16_t offset_y = 0;
+			
+			if (place_x != 0 && place_y != 0) {
+				offset_x = base_x - place_x;
+				offset_y = base_y - place_y;
+				base_x = base_x - offset_x;
+				base_y = base_y - offset_y;
+			}
```
Comment (gugahoa) [inline src/iomap.cpp]: This is the same as:
base_x = -place_x;
base_y = -place_y;

File: src/map.h [hunk 2]
```diff
@@ -185,8 +185,8 @@ class Map
 		  * Load a map.
 		  * \returns true if the map was loaded successfully
 		  */
-		bool loadMap(const std::string& identifier, bool loadHouses);
-
+		bool loadMap(const std::string& identifier, bool loadHouses, int place_x = 0, int place_y = 0);
+
```
Comment (gugahoa) [inline src/map.h]: Trailing whitespace

File: src/iomap.cpp [hunk 3]
```diff
@@ -198,9 +198,22 @@ bool IOMap::loadMap(Map* map, const std::string& identifier)
 				setLastErrorString("Invalid map node.");
 				return false;
 			}
-
-			uint16_t base_x = area_coord.x;
-			uint16_t base_y = area_coord.y;
+			
+			uint16_t base_x = 0;
```
Comment (djarek) [inline src/iomap.cpp]: This breaks if you load a map at a position that has multiple areas (all areas are loaded into the same coordinates).

**Comment by djarek** [comment]
Unfortunately this functionality isn't as simple as it might seem. This lua function is inherently unsafe - ***Bad Things™*** will hapen if a new map fragment overlays already existing tiles.

File: src/iomap.cpp (hunk 3 above)
Comment (urban119) [inline src/iomap.cpp]: no it no breaks. It is just need, becouse declarations must be assigned to anything. Default have 0.
In down code I am have "else", and that assignet to base_x = area_coord.x; (it be standard code).

**Comment by urban119** [comment]
It for me work. TFS use standard loadMap. If you in lua code write loadMap with attribute x,y only then You start use my modyfications.

File: src/iomap.cpp (hunk 3 above)
Comment (djarek) [inline src/iomap.cpp]: if we take this branch:
```
if (place_x != 0 && place_y != 0) {
```
then:
```
base_x = place_x; 
base_y=place_y;
```
which means that all areas within a map file will be loaded into the same coordinates(overlayed on top of each other).

**Comment by djarek** [comment]
Yeah it will work as long as the loaded map file has one area and the positions of the new tiles are empty. If not, the server might blow up.

File: src/iomap.cpp (hunk 1 above)
Comment (urban119) [inline src/iomap.cpp]: Yea you have right. I forgot remove that, becouse first my version been create offset.

File: src/iomap.cpp (hunk 3 above)
Comment (urban119) [inline src/iomap.cpp]: so, that is why is important to chose empty X,Y positions on Your map?

**Comment by urban119** [comment]
If you create script to load x5 map without calculate offset your width partMap... Yes you load x5 partMap on this same position and destroy u server.
You have two way... You clean all tiles x,y or just calculate offset -.^

**Comment by ranisalt** [comment]
Man, your phrases are completely unreadable

**Comment by djarek** [comment]
@urban119 
Okay, let's assume that the caller checks if the area that we load into is empty. This code still fails if you load in a map bigger than 255x255 tiles or with disjoint tile areas.

**Comment by urban119** [comment]
You are the better programmer in TFS, so if you have propositions, just write.
I create yourself OTS, where I am changing standard Quests for My "dynamics team style" raids, dungeons, wars etc. like a Word Of Warcaft.
I changed static loadMap to dynamic loadMap and published that changes for peoples. If someone use that.. its okey, if no so no.
That's all.

Regards
Urban

**Comment by ranisalt** [comment]
I'd like to point that not a single teleport will work if the map is offset.

**Comment by DSpeichert** [comment]
It looks like this PR would introduce a number of bugs as pointed out in review.

The TFS engine was not quite designed for dynamic map editing and this feature, while may be useful to you, does not seem to be something that is popular. If it worked, it could be reviewed but until then, I'm closing this PR.

**Comment by TheSumm** [comment]
Fixing the function to load the map properly would be fairly simple. It just requires to add the offsets to `area_coord`, instead of loading each map area at `place_x, place_y`.

Teleports are an issue though, since the destination might be relative to the loaded map (in this case no teleport would work properly) or the destination might be on the main map (in this case the teleport would work). However, a map might contain both kinds of teleports.

**Pull request closed without merging**